lib/handlers/events/delivery/goals/k8s/KubernetesGoalLauncher.ts

```diff
@@ -0,0 +1,329 @@
+            completedSdmJobs.map(j => `${j.metadata.namespace}:${j.metadata.name}`).join(", ")}`);
+
+        for (const completedSdmJob of completedSdmJobs) {
+            try {
+                await batch.deleteNamespacedJob(
+                    completedSdmJob.metadata.name,
+                    completedSdmJob.metadata.namespace,
+                    // propagationPolicy is needed so that pods of the job are also deleted
+                    { propagationPolicy: "Background" } as any);
+            } catch (e) {
+                logger.warn(`Failed to delete k8s goal job '${completedSdmJob.metadata.namespace}:${completedSdmJob.metadata.name}': ${e.message}`);
+            }
+        }
+    }
+}
+
+/**
+ * Create a k8s Job spec with affinity to jobs for the same goal set
+ * @param goalSetId
+ */
+function createJobSpecWithAffinity(podSpec: k8s.V1Pod, goalSetId: string): k8s.V1Job {
+    podSpec.spec.affinity = {
+        podAffinity: {
+            preferredDuringSchedulingIgnoredDuringExecution: [
+                {
+                    weight: 100,
+                    podAffinityTerm: {
+                        labelSelector: {
+                            matchExpressions: [
+                                {
+                                    key: "goalSetId",
+                                    operator: "In",
+                                    values: [
+                                        goalSetId,
+                                    ],
+                                },
+                            ],
+                        },
+                        topologyKey: "kubernetes.io/hostname",
+                    },
+                },
+            ],
+        },
+    } as any;
+
+    // Clean up podSpec
+    // See https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/#pod-v1-core note on nodeName
+    delete podSpec.spec.nodeName;
+
+    return {
+        kind: "Job",
+        apiVersion: "batch/v1",
+        metadata: {},
+        spec: {
+            template: {
+                metadata: {
+                    labels: {
+                        goalSetId,
+                    },
+                },
+                spec: podSpec.spec,
+            },
+        },
+    } as any;
+}
+
+/**
+ * Rewrite the volume host path to include the workspace id to prevent cross workspace content ending
+ * up in the same directory.
+ * @param deploymentSpec
+ * @param workspaceId
+ */
+function rewriteCachePath(deploymentSpec: k8s.V1Job, workspaceId: string): void {
+    const cachePath = configurationValue("sdm.cache.path", "/opt/data");
+    const containers: k8s.V1Container[] = _.get(deploymentSpec, "spec.template.spec.containers", []);
+
+    const cacheVolumeNames: string[] = [];
+    containers.forEach(c => {
+        cacheVolumeNames.push(...c.volumeMounts.filter(vm => vm.mountPath === cachePath).map(cm => cm.name));
+    });
+
+    _.uniq(cacheVolumeNames).forEach(vn => {
```

Comment:
Not a big deal, but it seems the previous forEach could have been a map and combined with this into a single statement.